### ISSUE TITLE
Define ExchangeType enum

### ISSUE THREAD

**Issue by jxramos** [pika/pika#1244]
It would be useful to have a dedicated enum which spells out all the fixed values available for python consumption much like you find in the .net RabbitMQ client

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/projects/client/RabbitMQ.Client/src/client/api/ExchangeType.cs

This source will likely be something like...

```
from enum import Enum

class ExchangeType(Enum) :
    direct = "Exchange type used for AMQP direct exchanges."
    fanout = "Exchange type used for AMQP fanout exchanges."
    headers = "Exchange type used for AMQP headers exchanges."
    topic = "Exchange type used for AMQP topic exchanges."
```

This will give folks the option to use enum references in their client code like

```
channel = self.connection.channel()
channel.exchange_declare(exchange='foo', exchange_type=ExchangeType.direct.name)
```

**Comment by lukebakken** [comment]
Hello and thanks for using Pika.

I would be happy to review a pull request implementing this. Thanks.

**Comment by jxramos** [comment]
I'll take it on this weekend. It would help me too to identify any other candidates for enums used throughout RabbitMQ. Enums are great for discovering siblings of a finite set of options straight for the sourcecode.

**Comment by deveshks** [comment]
Was this PR ever created and merged? If not, I will be happy to submit a PR for the same. I have used pika to a good extent in the past and would love to contribute.

**Comment by lukebakken** [comment]
This issue is still open, so no work has been done on it.

**Comment by deveshks** [comment]
Thanks @lukebakken for the response. Can I make a PR for this? Apart from the instruction in the README https://github.com/pika/pika#contributing , are there any other guidelines while making a PR?

**Comment by lukebakken** [comment]
Those guidelines are correct. There's no need to ask to open a PR, just do it 😄

**Comment by deveshks** [comment]
What would be a good place to define the Enum? In `channel.py`, or I can create a new module `exchange_type.py` ? `spec.py` looked to be a good place until I saw that it was auto generated.

Also I think in order to use the enum, the best way to go about it is what @jxramos  mentioned, which makes sure that the existing client code using pika library doesn't break.

```
channel = self.connection.channel()
channel.exchange_declare(exchange='foo', exchange_type=ExchangeType.direct.name)
````
 If we do something else like 
```channel.exchange_declare(exchange='foo', exchange_type=ExchangeType.direct)``` and then get the name within the code via 
```
return self._rpc(
            spec.Exchange.Declare(0, exchange, exchange_type.name, passive, durable,
                                  auto_delete, internal, nowait, arguments or
                                  dict()), callback,
            [spec.Exchange.DeclareOk] if not nowait else [])
```
that might break existing codebases using this repo. 

What do you think?

**Comment by lukebakken** [comment]
* Do whatever you think is most convenient for users of this library.
* Don't break backwards compatibility.
* Include tests.

Discussion can continue within the PR itself.

**Comment by seenureddy** [comment]
@deveshks @lukebakken Anyone worked on this issue?

**Comment by lukebakken** [comment]
Not yet.

**Comment by seenureddy** [comment]
@lukebakken I will work on this issue.

**Comment by seenureddy** [comment]
@lukebakken I have done the changes [https://github.com/seenureddy/pika/commit/d79ce1c1d576f1d39c72ef7f8aaeb48c3802302a](url). Need to test it.

**Comment by jxramos** [comment]
Thanks for picking this up @seenureddy, I got pulled off the tech stack working with RabbitMQ from my team but this is still desired by us. Having these literals tied to enums really improves the discoverability of the other relevant possibilities by simply looking up the enum's definition.

If there's anywhere else in the codebase that has scattered families of literals that could benefit from this consolidation that would make parameters and arguments more self documenting in their usage. Thanks again.

**Comment by seenureddy** [comment]
> 
> 
> It would be useful to have a dedicated enum which spells out all the fixed values available for python consumption much like you find in the .net RabbitMQ client
> 
> https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/projects/client/RabbitMQ.Client/src/client/api/ExchangeType.cs
> 
> This source will likely be something like...
> 
> ```
> from enum import Enum
> 
> class ExchangeType(Enum) :
>     direct = "Exchange type used for AMQP direct exchanges."
>     fanout = "Exchange type used for AMQP fanout exchanges."
>     headers = "Exchange type used for AMQP headers exchanges."
>     topic = "Exchange type used for AMQP topic exchanges."
> ```
> 
> This will give folks the option to use enum references in their client code like
> 
> ```
> channel = self.connection.channel()
> channel.exchange_declare(exchange='foo', exchange_type=ExchangeType.direct.name)
> ```

In my project I'm using the different exchange_type `EXCHANGE_TYPE = 'topic'`. Is it possible to use the same `ExchangeType` class for this type. Found in this location `examples/asynchronous_consumer_example.py`

**Comment by alecgerona** [comment]
@seenureddy is this already a PR?

**Comment by lukebakken** [comment]
#1260 will address this.

**Comment by lukebakken** [comment]
Implemented by #1260
